src/js_native_api_v8.cc

```diff
@@ -286,6 +285,17 @@ class Reference : private Finalizer {
   }
 
  private:
+  inline void SetWeak() {
+    _persistent.SetWeak(
+        this, FinalizeCallback, v8::WeakCallbackType::kParameter);
+    _env->track_finalizer(EnvCleanupHook, this);
```

Comment:
I’d go further and suggest the behaviour that Node.js itself uses (via `BaseObject`): Always call finalizers at the end of the current `Environment`, not just when they’re weak references.

Reply:
I think we have to work these kind of changes through the whole code base. For example, I still need to add this to `v8impl::CallbackBundle` as well. I'd like to make that a subclass of `Reference`.